client/mysqldump.c

```diff
@@ -4165,6 +4252,397 @@ static char *getTableName(int reset)
 } /* getTableName */
 
 
+/*
+  dump user/role grants
+*/
+
+static int dump_grants(const char *ur)
+{
+  DYNAMIC_STRING sqlbuf;
+  MYSQL_ROW row;
+  MYSQL_RES *tableres;
+
+  init_dynamic_string_checked(&sqlbuf, "SHOW GRANTS FOR ", 256, 1024);
+  dynstr_append_checked(&sqlbuf, ur);
+
+  if (mysql_query_with_error_report(mysql, &tableres, sqlbuf.str))
+  {
+    dynstr_free(&sqlbuf);
+    return 1;
+  }
+  while ((row= mysql_fetch_row(tableres)))
+  {
+    fprintf(md_result_file, "%s;\n", row[0]);
+  }
+  mysql_free_result(tableres);
+  dynstr_free(&sqlbuf);
+  return 0;
+}
+
+
+/*
+  dump creater user
+*/
+
+static int dump_create_user(const char *user)
+{
+  DYNAMIC_STRING sqlbuf;
+  MYSQL_ROW row;
+  MYSQL_RES *tableres;
+
+  init_dynamic_string_checked(&sqlbuf, "SHOW CREATE USER ", 256, 1024);
+  dynstr_append_checked(&sqlbuf, user);
+
+  if (mysql_query_with_error_report(mysql, &tableres, sqlbuf.str))
+  {
+    dynstr_free(&sqlbuf);
+    return 1;
+  }
+  while ((row= mysql_fetch_row(tableres)))
+  {
+    fprintf(md_result_file, "CREATE %sUSER %s%s;\n", opt_replace_into ? "/*M!100103 OR REPLACE */ ": "",
+            opt_ignore ? "IF NOT EXISTS " : "",
+            row[0] + sizeof("CREATE USER"));
+  }
+  mysql_free_result(tableres);
+  dynstr_free(&sqlbuf);
+  return 0;
+}
+
+
+
+/*
+  dump all users and roles
+*/
+
+static int dump_all_users()
+{
+  MYSQL_ROW row;
+  MYSQL_RES *tableres;
+  int result= 0;
+  /* Roles added in MariaDB-10.0.5 or MySQL-8.0 */
+  my_bool maria_roles_exist= (mysql_get_server_version(mysql) >= 100005);
+  my_bool mysql_roles_exist= (mysql_get_server_version(mysql) >= 80001) && !maria_roles_exist;
+
+  if (mysql_query_with_error_report(mysql, &tableres,
+                                    "SELECT CONCAT(QUOTE(u.user), '@', QUOTE(u.Host)) AS u "
+                                    "FROM mysql.user u "
+				    " /*!80001 LEFT JOIN mysql.role_edges e ON u.user=e.from_user AND u.host=e.from_host "
+				    "         WHERE e.from_user IS NULL */"
+				    " /*M!100005 WHERE is_role='N' */"))
+    return 1;
+  while ((row= mysql_fetch_row(tableres)))
+  {
+    if (opt_replace_into)
+      /* Protection against removing the current import user */
+      /* MySQL-8.0 export capability */
+      fprintf(md_result_file,
+	"DELIMITER |\n"
+        "/*M!100101 IF current_user()=\"%s\" THEN\n"
+        "  SIGNAL SQLSTATE '45000' SET MYSQL_ERRNO=30001, MESSAGE_TEXT=\"Don't remove current user %s'\";\n"
+        "END IF */|\n"
+	"DELIMITER ;\n"
+        "/*!50701 DROP USER IF EXISTS %s */;\n", row[0], row[0], row[0]);
+    if (dump_create_user(row[0]))
+      result= 1;
```

Comment:
So result is saying we have an error, but we are proceeding with a loop?

Reply:
I assumed the error was for a single user and might be transitory. There may be reasons that the next user may succeed. I know this isn't the greatest explanation, but trying to keep to a backup principle to be as thorough as possible, but still notify if errors. Is this ok?